### PR TITLE
[Week4] - huiju

### PR DESCRIPTION
**Pull request by momiju**

### CODE REVIEW

File: week4-huiju/task1.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        # TODO: 유닛 수가 64인 Dense 층을 추가하세요.
+        Dense(64),
+
+        # 해당 층에 배치 정규화와 ReLU를 추가합니다.
+        BatchNormalization(),
+        ReLU(),
+
+        # TODO: 유닛 수가 10인 출력층(Dense)과 softmax 활성화 함수를 추가하세요.
+        Dense(10),
+        Softmax()
```
Comment (kimhamyong) [inline week4-huiju/task1.py]: 출력층에서 Dense(10)과 별도로 Softmax()를 추가하는 것은 불필요합니다. Dense(10, activation='softmax')로 수정하여 간결하게 작성하는 것이 좋습니다.

```python
Dense(10, activation='softmax')
```

File: week4-huiju/task1.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    loss="sparse_categorical_crossentropy",
+    metrics=["accuracy"]
+)
+
+# TODO: 배치 크기를 64로, 에포크 수를 5로 설정하여 모델을 훈련하세요.
+model.fit(x_train, y_train, batch_size=64, epochs=5, verbose=1)
+
+
+# TODO: 모델을 평가하고 테스트 정확도를 출력하세요.
+print(f"Test Accuracy: {test_accuracy:.4f}")
```
Comment (kimhamyong) [inline week4-huiju/task1.py]: 이 부분 다시 수행해주세요!

File: week4-huiju/task1.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import tensorflow as tf
+from tensorflow.keras.models import Sequential
+from tensorflow.keras.layers import Dense, BatchNormalization, ReLU, Dropout, Softmax
+
+# 배치 정규화를 포함한 딥러닝 모델 정의 함수
+def create_model():
+    model = Sequential([
+        # TODO: 첫 번째 Dense 층(유닛 수: 256)과 적절한 입력 형태를 추가하세요.
+        Dense(256, input_shape=(784,)),
```
Comment (kimhamyong) [inline week4-huiju/task1.py:9]: 현재 케라스에선 Dense(256, input_shape=(784,)) 대신 Input 레이어를 사용하는 것이 더 권장되고 있습니다.

```python
from tensorflow.keras.layers import Input
model = Sequential([
    Input(shape=(784,)),
    Dense(256),
    ...
])

**Review by kimhamyong — DISMISSED** [review]
task 1:
모델 평가 결과에서 test_accuracy가 정의되지 않았습니다.

**Review by giyomdabin — CHANGES_REQUESTED** [review]
task2: 단계별 과제는 잘 수행하셨습니다!

다만, 주석 부분에 import 가 되어있어 오류가 발생하고 있습니다.

`# 필요한 라이브러리 가져오기import tensorflow as tf`

해당 부분에서 import의 위치를 변경해주세요! 수고하셨습니다.

**Comment by kimhamyong** [comment]
일반적으로 Dense -> BatchNormalization -> Activation 순서가 맞지만, 때에 따라 BatchNormalization -> Activation -> Dense를 선호하기도 합니다. 실험을 통해 최적의 성능을 확인하는 방법을 추천드립니다. 아래는 예시 코드입니다!

```python
def create_model_alternate_bn_order():
    model = Sequential([
        BatchNormalization(input_shape=(784,)),  # BatchNormalization 먼저 적용
        Dense(256),
        ReLU(),
        
        BatchNormalization(),
        Dense(128),
        ReLU(),
        Dropout(0.3),

        BatchNormalization(),
        Dense(64),
        ReLU(),
        Dropout(0.3),

        Dense(10, activation='softmax')
    ])
    return model

```

**Comment by kimhamyong** [comment]
MNIST 데이터셋의 정규화 범위가 (0, 1)로 설정되었습니다. 하지만 BatchNormalization을 사용하는 경우 데이터 정규화가 항상 필요하지 않을 수 있으므로 정규화를 제거한 버전에 대한 예시를 공유해드립니다.

```python
# 데이터 정규화를 제거합니다.
x_train_no_norm = x_train.reshape(-1, 784).astype('float32')
x_test_no_norm = x_test.reshape(-1, 784).astype('float32')

# 모델 생성 후 동일하게 훈련합니다.
model = create_model()
model.fit(
    x_train_no_norm, y_train,
    batch_size=64,
    epochs=5,
    validation_split=0.2
)

**Comment by kimhamyong** [comment]
- 학습 에포크 수를 5로 설정했는데, MNIST 데이터셋에서는 적절히 수렴하는지 확인이 필요합니다. 일반적으로 에포크를 더 늘리는 것이 유리할 수 있습니다.
- 배치 크기 64도 적절해 보이지만, 32 또는 128과 같은 다른 값도 시도해 볼 가치가 있습니다. 아래는 예시 코드입니다!

```python
# 배치 크기와 에포크 수를 조정하여 훈련
model.fit(
    x_train, y_train,
    batch_size=128,  # 배치 크기 증가
    epochs=10,       # 에포크 수 증가
    validation_split=0.2
)

**Comment by kimhamyong** [comment]
Dense(10, activation='softmax') 설정이 올바르지만, sparse_categorical_crossentropy를 사용할 때 Dense 층에서 activation='softmax'를 생략하더라도 내부적으로 자동으로 소프트맥스를 계산할 수 있습니다. 이를 명시적으로 언급하지 않는 코드 예제입니다.

```python
Dense(10)  # softmax 생략

**Comment by kimhamyong** [comment]
validation_split=0.2로 훈련 데이터의 일부를 검증 데이터로 사용했는데, 이를 별도의 검증 세트로 분리하면 학습 데이터를 더 잘 활용할 수 있습니다.

```python
from sklearn.model_selection import train_test_split
x_train_split, x_val, y_train_split, y_val = train_test_split(x_train, y_train, test_size=0.2, random_state=42)

model.fit(
    x_train_split, y_train_split,
    batch_size=64,
    epochs=5,
    validation_data=(x_val, y_val)
)

**Review by kimhamyong — APPROVED** [review]
task1: 잘 수행하셨습니다!
